components/Login/UserProfileForm.tsx

```diff
@@ -0,0 +1,368 @@
+import Head from 'next/head'
+import { useRouter } from 'next/router'
+import { useState, useRef } from 'react'
+import { auth, db, storage } from '../../firebase'
+import { createUserProfile } from '../../lib/db'
+import {
+    loginButtons,
+    loginDivs,
+    loginImages,
+    loginInputs,
+} from '../../styles/login'
+import Button from '../Utils/Button'
+import { Avatar, useMediaQuery } from '@mui/material'
+import { UilImagePlus, UilTrashAlt } from '@iconscout/react-unicons'
+import preventDefaultOnEnter from '../../hooks/preventDefaultOnEnter'
+import { useAuthState } from 'react-firebase-hooks/auth'
+
+type UserProfileFormProps = {
+    profile: firebase.firestore.DocumentData
+    closeModal: () => void
+}
+
+const UserProfileForm: React.FC<UserProfileFormProps> = ({
+    profile,
+    closeModal,
+}) => {
+    // User state
+    const [user] = useAuthState(auth)
+
+    // Form state
+    const [dm, setDm] = useState(profile.allowDM || false)
+    const [username, setUsername] = useState(profile.username || '')
+    const [name, setName] = useState('')
+    const [last, setLast] = useState('')
+    const [location, setLocation] = useState('')
+    const [bio, setBio] = useState('')
+    const [profilePic, setProfilePic] = useState('')
+
+    // Picture state
+    const profilePicRef = useRef(null)
+    const [imageToUpload, setImageToUpload] = useState(null)
+    const [targetEvent, setTargetEvent] = useState(null)
+
+    // Router
+    const router = useRouter()
+
+    // Handler functions
+    const toggleDM = () => {
+        setDm(!dm)
+    }
+
+    // Database Hook functions
+    const saveAndContinue = (e) => {
+        e.preventDefault()
+
+        // Upload profile pic
+        if (imageToUpload) {
+            // Push to storage
+            const uploadTask = storage
+                .ref(`profiles/${user.uid}`)
+                .putString(imageToUpload, 'data_url')
+
+            // Remove image preview
+            handleRemoveImage()
+
+            // When the state changes, add the url to profile
+            uploadTask.on(
+                'state_change',
+                null,
+                (error) => console.error(error),
+                () => {
+                    // When the upload completes
+                    storage
+                        .ref('profiles')
+                        .child(userProfile.uid)
+                        .getDownloadURL()
+                        .then((url) => {
+                            // Add the url to the user Profile
+                            // TODO: State not updating, possibly because async
+                            setProfilePic(url)
```

Comment:
Manually updating the `user` and `profile` collections here would probably do the trick. I tried avoid so as not to double to calls to the db...

Reply:
Ill experiment a couple of Firebase stuff.